src/QUICClient.ts

```diff
@@ -169,287 +156,399 @@ class QUICClient extends EventTarget {
-      );
-    } else if (
-      socket.type === 'ipv4' &&
-      utils.isIPv4MappedIPv6(socket.host) &&
-      !utils.isIPv4MappedIPv6(host_)
-    ) {
-      throw new errors.ErrorQUICClientInvalidHost(
-        `Cannot connect to ${host_} an IPv4 mapped IPv6 QUICClient`,
-      );
-    }
-    const abortController = new AbortController();
-    const abortHandler = () => {
-      abortController.abort(ctx.signal.reason);
-    };
-    ctx.signal.addEventListener('abort', abortHandler);
-    const connectionProm = QUICConnection.createQUICConnection(
-      {
-        type: 'client',
-        scid,
-        socket,
-        remoteInfo: {
-          host: host_,
-          port: port_,
-        },
-        config: quicConfig,
-        reasonToCode,
-        codeToReason,
-        verifyCallback,
-        logger: logger.getChild(
-          `${QUICConnection.name} ${scid.toString().slice(32)}`,
-        ),
-      },
-      ctx,
-    );
+
     try {
-      await Promise.race([connectionProm, socketErrorP]);
+      // Check that the target `host` is compatible with the bound socket host
+      if (
+        socket.type === 'ipv4' &&
+        !utils.isIPv4(host_) &&
+        !utils.isIPv4MappedIPv6(host_)
+      ) {
+        throw new errors.ErrorQUICClientInvalidHost(
+          `Cannot connect to ${host_} on an IPv4 QUICClient`,
+        );
+      } else if (
+        socket.type === 'ipv6' &&
+        (!utils.isIPv6(host_) || utils.isIPv4MappedIPv6(host_))
+      ) {
+        throw new errors.ErrorQUICClientInvalidHost(
+          `Cannot connect to ${host_} on an IPv6 QUICClient`,
+        );
+      } else if (socket.type === 'ipv4&ipv6' && !utils.isIPv6(host_)) {
+        throw new errors.ErrorQUICClientInvalidHost(
+          `Cannot send to ${host_} on a dual stack QUICClient`,
+        );
+      } else if (
+        socket.type === 'ipv4' &&
+        utils.isIPv4MappedIPv6(socket.host) &&
+        !utils.isIPv4MappedIPv6(host_)
+      ) {
+        throw new errors.ErrorQUICClientInvalidHost(
+          `Cannot connect to ${host_} an IPv4 mapped IPv6 QUICClient`,
+        );
+      }
     } catch (e) {
-      // In case the `connection.start` is ongoing, we need to abort it
-      abortController.abort(e);
       if (!isSocketShared) {
-        // Stop is idempotent
-        await socket.stop();
+        await socket.stop({ force: true });
       }
       throw e;
-    } finally {
-      socket.removeEventListener('socketError', handleQUICSocketError);
-      ctx.signal.removeEventListener('abort', abortHandler);
     }
-    const connection = await connectionProm;
-    address = utils.buildAddress(host_, port);
+    const connection = new QUICConnection({
+      type: 'client',
+      scid,
+      socket,
+      remoteInfo: {
+        host: host_,
+        port: port_,
+      },
+      config: quicConfig,
+      reasonToCode,
+      codeToReason,
+      logger: logger.getChild(`${QUICConnection.name} ${scid.toString()}`),
+    });
```

Comment:
Constructing the `QUICConnection` can throw an `ErrorQUICConnectionConfigInvalid` error. In this case we have a resource leak as we wont clean up the `QUICSocket`.

Reply:
Proceed to catch and stop the socket if it isn't shared.

Reply:
Done in a3bd8016662ef2b45ec10370dcd6e7e494f3ae77